pageserver/src/tenant/storage_layer.rs

```diff
@@ -81,6 +88,103 @@ pub enum ValueReconstructResult {
     Missing,
 }
 
+#[derive(Default)]
+pub struct LayerAccessStats(RwLock<LayerAccessStatsInner>);
+
+#[derive(Clone)]
+struct LayerAccessStatFullDetails {
+    when: SystemTime,
+    task_kind: TaskKind,
+    access_kind: LayerAccessKind,
+}
+
+#[derive(Default)]
+struct LayerAccessStatsInner {
+    first_access: Option<LayerAccessStatFullDetails>,
+    count_by_access_kind: EnumMap<LayerAccessKind, u64>,
+    task_kind_flag: EnumSet<TaskKind>,
+    last_accesses: VecDeque<LayerAccessStatFullDetails>,
+}
+
+#[derive(Clone, Copy, strum_macros::EnumString)]
+pub enum LayerAccessStatsReset {
+    JustTaskKindFlags,
+    AllStats,
+}
+
+impl LayerAccessStatFullDetails {
+    fn to_api_model(&self) -> pageserver_api::models::LayerAccessStatFullDetails {
+        let Self {
+            when,
+            task_kind,
+            access_kind,
+        } = self;
+        pageserver_api::models::LayerAccessStatFullDetails {
+            when_millis_since_epoch: when
+                .duration_since(UNIX_EPOCH)
+                .expect("better to die in this unlikely case than report false stats")
+                .as_millis(),
+            task_kind: task_kind.into(), // into static str, powered by strum_macros
+            access_kind: *access_kind,
+        }
+    }
+}
+
+pub static LAYER_ACCESS_STATS_KILLSWITCH: AtomicBool = AtomicBool::new(false);
+
+impl LayerAccessStats {
+    fn record_access(&self, access_kind: LayerAccessKind, task_kind: TaskKind) {
+        if LAYER_ACCESS_STATS_KILLSWITCH.load(atomic::Ordering::SeqCst) {
```

Comment:
What's the idea behind the killswitch here only?
Other methods are still accessing the lock without the switch, should they consider it too?

Reply:
This function is the only one called from the data path.
The others need not be considered since they are only called from the new HTTP endpoints.

(I added this killswitch from the perspective of the potential perf impact of this PR, not "could it crash".)

Reply:
The SeqCst global order is giving me headache.. Why is it necessary here instead of Relaxed?

Reply:
Yeah, due to the subsequent `.lock()`, `Relaxed` would be fine.

Reply:
Decided to remove the killswitch, will mark this as resolved.